tests/unit/test_matrix.cu

```diff
@@ -2,64 +2,386 @@
+        std::vector<I> starts,
+        std::vector<T> values,
+        int padded_size)
+{
+    auto num_mtx = sizes.size();
+
+    auto in  = on_gpu(memory::make_const_view(values));
+    auto sizes_d = on_gpu(memory::make_const_view(sizes));
+    auto starts_d = on_gpu(memory::make_const_view(starts));
+
+    int packed_size = padded_size * BlockWidth * gpu::impl::block_count(num_mtx, BlockWidth);
+
+    // forward will hold the result of the interleave operation on the GPU
+    auto forward = memory::device_vector<T>(packed_size, npos<T>());
+
+    // find the reference interleaved values using host side implementation
+    auto baseline = gpu::flat_to_interleaved(values, sizes, starts, BlockWidth, num_mtx, padded_size);
+
+    // find the interleaved values on gpu
+    gpu::flat_to_interleaved<T, I, BlockWidth, LoadWidth>(in.data(), forward.data(), sizes_d.data(), starts_d.data(), padded_size, num_mtx);
+
+    std::vector<T> result_f = assign_from(on_host(forward));
+    std::vector<T> expected = gpu::flat_to_interleaved(values, sizes, starts, BlockWidth, num_mtx, padded_size);
+    const auto forward_success = (result_f==expected);
+    if (!forward_success) {
+        return ::testing::AssertionFailure() << "interleave to flat failed: BlockWidth "
+            << BlockWidth << ", LoadWidth " << LoadWidth << "\n";
+    }
+
+    // backward will hold the result of reverse interleave on the GPU
+    auto backward = memory::device_vector<T>(values.size(), npos<T>());
+    gpu::interleaved_to_flat<T, I, BlockWidth, LoadWidth>(forward.data(), backward.data(), sizes_d.data(), starts_d.data(), padded_size, num_mtx);
+
+    std::vector<T> result_b = assign_from(on_host(backward));
+
+    // we expect that the result of the reverse permutation is the original input vector
+    const auto backward_success = (result_b==values);
+    EXPECT_TRUE(backward_success);
+    if (!backward_success) {
+        return ::testing::AssertionFailure() << "flat to interleave failed: BlockWidth "
+            << BlockWidth << ", LoadWidth " << LoadWidth << "\n";
+    }
+
+    return ::testing::AssertionSuccess();
+}
+
+// test conversion to and from interleaved back end storage format
+TEST(matrix, interleave)
 {
-    using namespace nest::mc;
+    using I = int;
+    using T = int;
+    using ivec = std::vector<I>;
+    using tvec = std::vector<T>;
+
+    // simple case with 4 matrices of length 2
+    {
+        const int padded_size = 2;
+        const int num_mtx = 4;
+        ivec sizes(num_mtx, padded_size);
+
+        // find the start position of each matrix in the flat storage
+        // we are assuming that the matrices are unpermuted
+        ivec starts(num_mtx, 0);
+        std::partial_sum(begin(sizes), end(sizes)-1, begin(starts)+1);
+
+        tvec values(padded_size*num_mtx);
+        std::iota(values.begin(), values.end(), 0);
 
-    using nest::mc::util::make_span;
+        test_interleave<T, I, 1, 1>(sizes, starts, values, padded_size);
```

Comment:
The assertion results from `test_interleave` calls need to be tested with `EXPECT_TRUE` or similar.